central/delegatedregistryconfig/delegator/delegator.go

```diff
@@ -0,0 +1,148 @@
+package delegator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/stackrox/rox/central/delegatedregistryconfig/datastore"
+	"github.com/stackrox/rox/central/sensor/service/connection"
+	"github.com/stackrox/rox/generated/internalapi/central"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/centralsensor"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/urlfmt"
+	"github.com/stackrox/rox/pkg/waiter"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+// New creates a new delegator.
+func New(deleRegConfigDS datastore.DataStore, connManager connection.Manager, scanWaiterManager waiter.Manager[*storage.Image]) *delegatorImpl {
+	return &delegatorImpl{
+		deleRegConfigDS:   deleRegConfigDS,
+		connManager:       connManager,
+		scanWaiterManager: scanWaiterManager,
+	}
+}
+
+type delegatorImpl struct {
+	// deleRegConfigDS for pulling the current delegated registry config.
+	deleRegConfigDS datastore.DataStore
+
+	// connManager for sending scan requests to secured clusters and ensuring
+	// clusters are valid for delegation.
+	connManager connection.Manager
+
+	// scanWaiterManager creates waiters that wait for async scan responses.
+	scanWaiterManager waiter.Manager[*storage.Image]
+}
+
+// GetDelegateClusterID returns the cluster id that should enrich this image (if any) and
+// true if enrichment should be delegated to a secured cluster, false otherwise.
+func (d *delegatorImpl) GetDelegateClusterID(ctx context.Context, image *storage.Image) (string, bool, error) {
+	config, err := d.getConfig(ctx)
+	if err != nil {
+		return "", false, err
+	}
+
+	shouldDelegate, clusterID := d.shouldDelegate(image, config)
+	if !shouldDelegate {
+		return "", false, nil
+	}
+
+	err = d.validateCluster(clusterID)
+	return clusterID, true, err
+}
+
+// DelegateEnrichImage sends an enrichment request to the provided cluster.
+func (d *delegatorImpl) DelegateEnrichImage(ctx context.Context, image *storage.Image, clusterID string, force bool) error {
+	if clusterID == "" {
+		return errors.New("missing cluster id")
+	}
+
+	w, err := d.scanWaiterManager.NewWaiter()
+	if err != nil {
+		return err
+	}
+
+	msg := &central.MsgToSensor{
+		Msg: &central.MsgToSensor_ScanImage{
+			ScanImage: &central.ScanImage{
+				RequestId: w.ID(),
+				ImageName: image.GetName().GetFullName(),
+				Force:     force,
+			},
+		},
+	}
+
+	err = d.connManager.SendMessage(clusterID, msg)
+	if err != nil {
+		w.Close()
+		return err
+	}
+
+	log.Infof("Sent scan request %q to cluster %q for %q", w.ID(), clusterID, image.GetName().GetFullName())
+
+	img, err := w.Wait(ctx)
+	if err != nil {
+		return fmt.Errorf("error delegating scan to cluster %q for %q: %w", clusterID, image.GetName().GetFullName(), err)
+	}
+
+	log.Debugf("Scan response received for %q and image %q", w.ID(), img.GetName().GetFullName())
+
+	// Copy the fields from img into image, callers expecting image to be modified in place.
+	*image = *img
```

Comment:
Why do we expect this to modified in place? It seems cleaner to return an image that is enriched

Reply:
The methods in `pkg/images/enricher/enricher_impl.go` that "enrich" modify in place, and the callers of those methods expect the update in place.  I agree a return is cleaner, in this case was following the convention.

I'll change the `Delegator` to return the image and rename the func to `DelegateScanImage` to keep the convention, the update in place will still need to occur, will be moved up the stack in this case.